### PR TITLE
Add shelving to NiftiMasker

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Supersed #994

### CODE REVIEW

File: nilearn/_utils/cache_mixin.py [hunk 1]
```diff
@@ -117,6 +119,10 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
         be cached or not (if user_memory_level is equal of greater than
         func_memory_level the function is cached)
 
+    shelve: boolean,
+        Whether to return a MemorizedResult, callable by a .get() method,
```
Comment (GaelVaroquaux) [inline nilearn/_utils/cache_mixin.py]: Say that it's a joblib MemorizedResult, so that people who don't know what it is can Google.

File: nilearn/_utils/cache_mixin.py [hunk 2]
```diff
@@ -189,6 +209,10 @@ def _cache(self, func, func_memory_level=1, **kwargs):
             The memory_level from which caching must be enabled for the wrapped
             function.
 
+        shelve: boolean,
+            Whether to return a MemorizedResult, callable by a .get() method,
```
Comment (GaelVaroquaux) [inline nilearn/_utils/cache_mixin.py]: Same remark

File: nilearn/input_data/multi_nifti_masker.py [hunk 3]
```diff
@@ -267,16 +270,19 @@ def transform_imgs(self, imgs_list, confounds=None, copy=True, n_jobs=1):
...
-                           memory=self.memory,
-                           verbose=self.verbose,
-                           confounds=cfs,
-                           copy=copy)
+                          memory_level=self.memory_level,
+                          memory=self.memory,
+                          verbose=self.verbose,
+                          confounds=cfs,
+                          copy=copy,
+                          return_affine=False)
```
Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_masker.py]: While I agree that all these cosmetic changes make the code abide better to standards, they make reviewing the PR hard. :(

File: nilearn/input_data/multi_nifti_masker.py [hunk 4]
```diff
@@ -267,16 +270,19 @@ def transform_imgs(self, imgs_list, confounds=None, copy=True, n_jobs=1):
...
             for imgs, cfs in izip(niimg_iter, confounds))
-        return [d[0] for d in data]
+        return data
```
Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_masker.py]: Looking at this, it seems that the signature of the method has changed. Is that the case?

File: nilearn/input_data/nifti_masker.py [hunk 5]
```diff
@@ -31,8 +31,8 @@ def filter_and_mask(imgs, mask_img_, parameters,
                     memory_level=0, memory=Memory(cachedir=None),
                     verbose=0,
                     confounds=None,
-                    copy=True):
-
+                    copy=True,
+                    return_affine=True):
```
Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_masker.py]: Why this change? It add an argument that may trigger recomputation and doesn't add much to the function.

File: nilearn/_utils/cache_mixin.py (hunk 1 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: Done.

File: nilearn/_utils/cache_mixin.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: Done.

File: nilearn/input_data/nifti_masker.py [hunk 6]
```diff
@@ -189,6 +192,24 @@ def __init__(self, mask_img=None, sessions=None, smoothing_fwhm=None,
...
+
+    def _activate_shelving(self):
+        """Activate shelving.
+
+        When shelving is activated, the transform method will shelve the result
+        on disk and return an object region_signals with method get:
+        region_signals.get() yields the maked array.
+        Useful to parallelize the masking of records before entering a
+        sequential pipeline."""
+        self._shelving = True
```
Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_masker.py]: Why the two methods above. They are on-liners. They should be removed, IMHO.

File: nilearn/input_data/nifti_masker.py [hunk 7]
```diff
@@ -189,6 +192,24 @@ def __init__(self, mask_img=None, sessions=None, smoothing_fwhm=None,
...
+
+    def _toggle_shelving(self):
+        self._shelving = not self._shelving
```
Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_masker.py]: I am not really that this method is useful either.

File: nilearn/input_data/multi_nifti_masker.py (hunk 4 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: No because `return_affine=False`. See below.

Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_masker.py]: > No because return_affine=False. See below.

I saw that after, but I don't like it :)

File: nilearn/input_data/nifti_masker.py (hunk 5 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_masker.py]: The thing is that `filter_and_mask` returns the affine because we need it to do sanity checks (reminder, for performance reasons, the `check_niimg` is called in `filter_and_mask`). So, if we directly shelve the result of `filter_and_mask`, the affine will be stored together, which is not useful. Plus, if we need the affine, we will have to unshelve the object.

One solution may be to add a `shelving` parameter to `filter_and_mask` and directly use the shelving inside.

File: nilearn/input_data/nifti_masker.py (hunk 6 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_masker.py]: I agree. We only need a `_set_shelving` one.

Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_masker.py]: > I agree. We only need a _set_shelving one.

We don't need a setter. This is not java. We just change the value of the
attribute directly. The methods do nothing else.

File: nilearn/_utils/cache_mixin.py [hunk 8]
```diff
@@ -117,6 +119,10 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
         be cached or not (if user_memory_level is equal of greater than
         func_memory_level the function is cached)
 
+    shelve: boolean,
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: No comma needed I think. Also I tend to prefer Python types in general so bool rather than boolean.

File: nilearn/_utils/cache_mixin.py [hunk 9]
```diff
@@ -157,7 +163,23 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
...
+    cached_func = _safe_cache(memory, func, **kwargs)
+    if shelve:
+        if LooseVersion(sklearn.__version__) < LooseVersion('0.15'):
+            raise ValueError('Shelving is only available if'
+                             ' scikit-learn >= 0.15 is installed.')
+        cached_func = _shelved_func(cached_func)
+    return cached_func
+
+
+class _shelved_func(object):
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: _ShelvedFunc would abide by the usual naming conventions for classes.

File: nilearn/_utils/cache_mixin.py [hunk 10]
```diff
@@ -189,6 +211,10 @@ def _cache(self, func, func_memory_level=1, **kwargs):
             The memory_level from which caching must be enabled for the wrapped
             function.
 
+        shelve: boolean,
```
Comment (lesteve) [inline nilearn/_utils/cache_mixin.py]: Same comment as above.

File: nilearn/input_data/multi_nifti_masker.py [hunk 11]
```diff
@@ -4,20 +4,20 @@
...
 import itertools
+import warnings
 
 from sklearn.externals.joblib import Memory, Parallel, delayed
 
-from .. import masking
-from .. import image
-from .. import _utils
-from .._utils import CacheMixin
+from nilearn import _utils
```
Comment (lesteve) [inline nilearn/input_data/multi_nifti_masker.py]: We should use relative imports everywhere ...

**Comment by bthirion** [comment]
This looks good, but shouldn't shelving appear somewhere in the docs (maybe simply as a reference to joblib ?). If you don't want to bother neuroimagers with it and keep this feature for experts (if I read correctly, this is deactivated by default), please include a few comments/references in the code to help people understand what is done.

**Comment by GaelVaroquaux** [comment]
> This looks good, but shouldn't shelving appear somewhere in the docs
> (maybe simply as a reference to joblib ?).

I think that so far we don't want to expose it to end users. It will be
used by the MSDL implementation.

File: nilearn/input_data/multi_nifti_masker.py (hunk 4 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/multi_nifti_masker.py]: Fixed it by removing the kwarg.

**Comment by AlexandreAbraham** [comment]
Comments addressed, should be good to go.

**Comment by AlexandreAbraham** [comment]
I addressed all the comments so this should be ready to go? @GaelVaroquaux

File: nilearn/input_data/nifti_masker.py [hunk 12]
```diff
@@ -268,8 +269,9 @@ def transform_single_imgs(self, imgs, confounds=None, copy=True):
 
         Returns
         -------
-        region_signals: 2D numpy.ndarray
-            Signal for each voxel inside the mask.
+        region_signals: 2D numpy.ndarray or MemorizedResult
```
Comment (KamalakerDadi) [inline nilearn/input_data/nifti_masker.py]: May be we should forget about ```MemorizedResult``` since it is not exposed to user and also it is not officially documented yet.

File: nilearn/_utils/cache_mixin.py [hunk 13]
```diff
@@ -157,7 +173,13 @@ def cache(func, memory, func_memory_level=None, memory_level=None,
...
     else:
         memory = Memory(cachedir=None, verbose=verbose)
-    return _safe_cache(memory, func, **kwargs)
+    cached_func = _safe_cache(memory, func, **kwargs)
+    if shelve:
+        if LooseVersion(sklearn.__version__) < LooseVersion('0.15'):
+            raise ValueError('Shelving is only available if'
+                             ' scikit-learn >= 0.15 is installed.')
+        cached_func = _ShelvedFunc(cached_func)
+    return cached_func
```
Comment (KamalakerDadi) [inline nilearn/_utils/cache_mixin.py]: Do we required to say in documentation what it returns if it is shelved ?

File: nilearn/_utils/cache_mixin.py [hunk 14]
```diff
@@ -251,4 +276,5 @@ def _cache(self, func, func_memory_level=1, **kwargs):
             self.memory_level = 1
 
         return cache(func, self.memory, func_memory_level=func_memory_level,
-                     memory_level=self.memory_level, **kwargs)
+                     memory_level=self.memory_level, shelve=shelve,
+                     **kwargs)
```
Comment (KamalakerDadi) [inline nilearn/_utils/cache_mixin.py]: same comment here as above ?

**Comment by KamalakerDadi** [comment]
Some minor comments from my side.

File: nilearn/_utils/cache_mixin.py (hunk 13 above)
Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: You mean the hack regarding _ShelvedFunc?

File: nilearn/input_data/nifti_masker.py (hunk 12 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_masker.py]: Indeed.

File: nilearn/_utils/cache_mixin.py (hunk 13 above)
Comment (KamalakerDadi) [inline nilearn/_utils/cache_mixin.py]: Yes. Is it not required ?

Comment (AlexandreAbraham) [inline nilearn/_utils/cache_mixin.py]: I modified the doc!

**Comment by KamalakerDadi** [comment]
LGTM

**Comment by KamalakerDadi** [comment]
Actually, confused with ```memory_level``` and ```func_memory_level``` in ```_cache()``` function ?

Isn't that should be called as ```memory_level``` ?

**Comment by AlexandreAbraham** [comment]
> Actually, confused with memory_level and func_memory_level in _cache() function ?

Oh, this... You must know about memory_level, right? Well, this function will decide if the caching is activated or not.

`func_memory_level` is the `threshold` from which the caching must be activated. `memory_level` is the level selected by the used. If `memory_level` is above `func_memory_level` (or the opposite, I always mix it up), then caching is activated, otherwise it is not.

I think that we should remove that but, for the moment, this is how it works.
